### PR TITLE
Handle AssemblyResolve events when the requesting assembly is unknown

### LINKED ISSUE #635: AnalyzerFileReference.InMemoryAssemblyLoader should always try to satisfy AssemblyResolve events
When we load an analyzer assembly A we also hook `AppDomain.CurrentDomain.AssemblyResolve` so that we can handle loading A's dependencies. When our event hook is called we inspect `ResolveEventArgs.RequestingAssembly` and only try to handle the event if the requesting assembly is an analyzer.

In practice we've discovered that `RequestingAssembly` is sometimes null, leading us to miss our chance to load a dependency.

Instead, we should always attempt to handle the `AssemblyResolve` event and only use `RequestingAssembly` (when available) as a hint for where to look first.

### PR DESCRIPTION
**Pull request by tmeschter**
Since we handle loading analyzer assemblies ourselves (to avoid locking
the file on disk and the need to shadow copy them) we must also hook the
AssemblyResolve event and handle finding and loading their dependencies.
Currently we only handle the event if the event arguments specify the
requesting assembly (via `ResolveEventArgs.RequestingAssembly`) and the
requesting assembly is known to us, i.e. we were responsible for loading
it.

Unfortunately `ResolveEventArgs.RequestingAssembly` is more of a hint
than anything else, and there are circumstances where it can be null. As
such we sometimes fail to load a dependency needed by an analyzer, and
it doesn't run.

The fix here is to attempt to load the assembly even when
`RequestingAssembly` is null. Instead of looking next to the requesting
assembly, we will look next to every assembly we have already loaded for
one with the correct identity, and load and return that.

Note that this change does *not* attempt to deal with cross-platform issues;
i.e., how do we handle loading analyzer assemblies in CoreCLR. That will
come later.

Fixes #635.

### CODE REVIEW

**Comment by tmeschter** [comment]
Ping @srivatsn @JohnHamby @shyamnamboodiripad @mavasani @jmarolf @heejaechang

File: src/Compilers/Core/Desktop/AnalyzerFileReference.InMemoryAssemblyLoader.cs [hunk 1]
```diff
@@ -193,6 +255,55 @@ private static Assembly CurrentDomain_AssemblyResolve(object sender, ResolveEven
                     return assembly;
                 }
             }
+
+            private static AssemblyIdentity TryGetAssemblyIdentity(string filePath)
+            {
+                try
+                {
+                    using (var stream = new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read | FileShare.Delete))
```
Comment (mavasani) [inline src/Compilers/Core/Desktop/AnalyzerFileReference.InMemoryAssemblyLoader.cs]: Probably worth adding a check for File.Exists before trying to open it here? Seems like we will call it for all combinations of "<LoadedAssemblyDir>.RequestedAssemblyName", lot of which are non-existent paths...

File: src/Compilers/Core/Desktop/AnalyzerFileReference.InMemoryAssemblyLoader.cs [hunk 2]
```diff
@@ -143,23 +146,82 @@ private static Assembly LoadCore(string fullPath)
...
-            /// We assume that an assembly's dependencies can be found next to it in the file
-            /// system.
+            /// Handles the <see cref="AppDomain.AssemblyResolve"/> event.
             /// </summary>
             private static Assembly CurrentDomain_AssemblyResolve(object sender, ResolveEventArgs args)
             {
                 if (args.RequestingAssembly == null)
                 {
-                    // We don't know who is requesting the load; don't try to satisfy the request.
+                    return ResolveForUnknownRequestor(args.Name);
```
Comment (heejaechang) [inline src/Compilers/Core/Desktop/AnalyzerFileReference.InMemoryAssemblyLoader.cs]: just  curious when dll will be actually loaded by CLR? is that only place where dlls will be loaded? asking as perf champ curious about VM impact.

File: src/Compilers/Core/Desktop/AnalyzerFileReference.InMemoryAssemblyLoader.cs (hunk 1 above)
Comment (tmeschter) [inline src/Compilers/Core/Desktop/AnalyzerFileReference.InMemoryAssemblyLoader.cs]: Good point.

File: src/Compilers/Core/Desktop/AnalyzerFileReference.InMemoryAssemblyLoader.cs (hunk 2 above)
Comment (tmeschter) [inline src/Compilers/Core/Desktop/AnalyzerFileReference.InMemoryAssemblyLoader.cs]: The CLR doesn't load the assembly. Instead, we load it when we call `Assembly.Load(byte[])` (see the `LoadCore` method).

Comment (tmeschter) [inline src/Compilers/Core/Desktop/AnalyzerFileReference.InMemoryAssemblyLoader.cs]: I would not expect this change to have a VM impact. All it does is ensure we load an `Assembly` that we should be loading anyway; we don't load anything that we don't actually need.

File: src/Compilers/Core/Desktop/AnalyzerFileReference.InMemoryAssemblyLoader.cs [hunk 3]
```diff
@@ -143,23 +146,82 @@ private static Assembly LoadCore(string fullPath)
...
+            private static Assembly ResolveForUnknownRequestor(string requestedAssemblyName)
+            {
+                lock (s_guard)
+                {
+                    string requestedNameWithPolicyApplied = AppDomain.CurrentDomain.ApplyPolicy(requestedAssemblyName);
+
+                    Assembly assembly;
+                    if (s_assembliesFromNames.TryGetValue(requestedNameWithPolicyApplied, out assembly))
+                    {
+                        // We've already loaded an assembly by this name; use that.
```
Comment (shyamnamboodiripad) [inline src/Compilers/Core/Desktop/AnalyzerFileReference.InMemoryAssemblyLoader.cs]: Does requestedNameWithPolicyApplied already include assembly identity info? Is it safe to just return "any" assembly that matches the supplied name?

Comment (shyamnamboodiripad) [inline src/Compilers/Core/Desktop/AnalyzerFileReference.InMemoryAssemblyLoader.cs]: Ah never mind. Based on following line looks like it does include identify info...

**Comment by jmarolf** [comment]
:+1:

**Comment by srivatsn** [comment]
:+1:

**Comment by mavasani** [comment]
:+1:

**Pull request merged**